sql/core/src/main/scala/org/apache/spark/sql/execution/DataSourceScanExec.scala

```diff
@@ -167,28 +167,30 @@ case class FileSourceScanExec(
       partitionSchema = relation.partitionSchema,
       relation.sparkSession.sessionState.conf)
 
-  val driverMetrics: HashMap[String, Long] = HashMap.empty
-
   /**
    * Send the driver-side metrics. Before calling this function, selectedPartitions has
    * been initialized. See SPARK-26327 for more details.
    */
   private def sendDriverMetrics(): Unit = {
+    val driverMetrics: HashMap[String, Long] = HashMap.empty
+
+    driverMetrics("numFiles") = selectedPartitions.map(_.files.size.toLong).sum
+    val fileListingPhase = relation.location.fileListingPhase
+    if (fileListingPhase.isDefined) {
+      val phase = fileListingPhase.get
+      driverMetrics("metadataTime") = phase.durationMs
```

Comment:
As discussion with @gatorsmile, `metadataTime` maybe still a clearer name for users, change it back in b79e948.
As my current plan, we'll finally get five metrics as below:
"metadata time": including all metadata accessing time like file listing time, meta store time and others.
"file listing start/end": the start and end timestamp for file listing
"metastore operation start/end": the start and end timestamp for metastore operation.